Add validation and operations tests for clearBuffer

toji opened the pull request:
<hr>

**Author checklist for test code/plans:**

- [x] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [x] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

kainino0x commented on src/webgpu/api/operation/command_buffer/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,54 @@
+      .combine('size', [0, 4, 8, 16, undefined])
+      .expand('bufferSize', p => [
+        (p.offset || 0) + (p.size || 16),
+        (p.offset || 0) + (p.size || 16) + 8,
```
nit: Prefer the `??` ("nullish coalescing") operator over `||` for this

kainino0x commented on src/webgpu/api/operation/command_buffer/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,54 @@
+    t.device.queue.submit([encoder.finish()]);
+
+    const expectOffset = offset === undefined ? 0 : offset;
+    const expectSize = size === undefined ? bufferSize - expectOffset : size;
```
`??` should work here too

kainino0x commented on src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,196 @@
+  });
+
+g.test('default_args')
+  .desc(`Test that calling clearBuffer with a default offset and size is valid.`)
```
Should also test with offset=undefined,size=value; and offset=value,size=undefined

kainino0x commented on src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,196 @@
+    Test that the clear size must be 4 byte aligned.
+    - Test size is not a multiple of 4.
+    - Test size is 0.
+    - Test size overflows the buffer size.
```
I don't see a subcase for this. Looks like it's probably tested in `'overflow'` anyway.

kainino0x commented on src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,196 @@
+    `
+    Test that the clear offsets must be 4 byte aligned.
+    - Test offset is not a multiple of 4.
+    - Test offset is larger than the buffer size.
```
same

kainino0x commented on src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,196 @@
+  });
+
+g.test('out_of_bounds')
+  .desc(`Test that clears which exceed the buffer bounds invalid.`)
```
nit:
```suggestion
  .desc(`Test that clears which exceed the buffer bounds are invalid.`)
```

kainino0x commented on src/webgpu/api/validation/encoding/cmds/clearBuffer.spec.ts:
```diff
@@ -0,0 +1,196 @@
+  .desc(`Test that clears which exceed the buffer bounds invalid.`)
+  .paramsSubcasesOnly([
+    { offset: 0, size: 32, _isSuccess: true },
+    { offset: 0, size: 36 },
```
Should probably test edge cases like `{ offset: 32, size: 0 }, { offset: 32, size: 4 }`

kainino0x reviewed (approved):
LGTM with nits

kainino0x commented on package.json:
```diff
@@ -40,7 +40,7 @@
     "@types/serve-index": "^1.9.1",
     "@typescript-eslint/parser": "^4.22.0",
-    "@webgpu/types": "^0.1.7",
+    "@webgpu/types": "^0.1.8",
```
Sorry, I caused a merge conflict in #848. Will need to change this to `0.1.8` (without the caret `^`).

toji replied:
Actually, I've found that there's an error in 0.1.8. Fix incoming. Also, there's a few changes that need to be made to the CTS to roll to the new types so I think I'll handle that as a separate PR.

toji commented:
Addressed feedback, but this PR is dependent of https://github.com/gpuweb/cts/pull/849, which is dependent on https://github.com/gpuweb/types/pull/87, so it's not going to clear the checks yet.
